securedrop_client/logic.py

```diff
@@ -574,57 +575,114 @@ def on_reply_download_failure(self, exception: Exception) -> None:
-        submission_filepath = os.path.join(self.data_dir, fn_no_ext)
-        original_filepath = os.path.join(self.data_dir, file.original_filename)
-
-        if os.path.exists(original_filepath):
-            os.remove(original_filepath)
-        os.link(submission_filepath, original_filepath)
-        if self.proxy or self.qubes:
-            # Running on Qubes.
-            command = "qvm-open-in-vm"
-            args = ['$dispvm:sd-svs-disp', original_filepath]
-
-            # QProcess (Qt) or Python's subprocess? Who cares? They do the
-            # same thing. :-)
-            process = QProcess(self)
-            process.start(command, args)
-        else:  # pragma: no cover
-            # Non Qubes OS. Just log the event for now.
-            logger.info('Opening file "{}".'.format(original_filepath))
-
-    def run_export_preflight_checks(self):
+        logger.info('Opening file "{}".'.format(file.original_filename))
+
+        fn_no_ext, dummy = os.path.splitext(os.path.splitext(file.filename)[0])
+        filepath = os.path.join(self.data_dir, fn_no_ext)
+        if not os.path.exists(filepath):
+            msg = _('Could not open {}. File does not exist.'.format(file.original_filename))
+            storage.mark_as_not_downloaded(file_uuid, self.session)
+            self.sync_api()
+            logger.debug(msg)
+            self.gui.update_error_status(msg)
+            return
+
+        path_to_file_with_original_name = os.path.join(self.data_dir, file.original_filename)
+        if not os.path.exists(path_to_file_with_original_name):
+            os.link(filepath, path_to_file_with_original_name)
+
+        if not self.qubes:
+            return
+
+        command = "qvm-open-in-vm"
+        args = ['$dispvm:sd-svs-disp', path_to_file_with_original_name]
+        process = QProcess(self)
+        process.start(command, args)
+
+    def run_export_preflight_checks(self, file_uuid: str):
         '''
-        Run preflight checks to make sure the Export VM is configured correctly
+        Run preflight checks to make sure the Export VM is configured correctly. Also check if the
+        file we're about to export exists before asking the user to connect their thumb drive and
+        type in their passphrase to unlock it.
         '''
-        logger.debug('Calling export preflight checks from thread {}'.format(
-            threading.current_thread().ident))
+        logger.info('Running export preflight checks')
+
+        file = self.get_file(file_uuid)
+        fn_no_ext, dummy = os.path.splitext(os.path.splitext(file.filename)[0])
+        filepath = os.path.join(self.data_dir, fn_no_ext)
+        if not os.path.exists(filepath):
+            msg = _('Could not export {}. File does not exist.'.format(file.original_filename))
+            storage.mark_as_not_downloaded(file_uuid, self.session)
+            self.sync_api()
+            logger.debug(msg)
+            self.gui.update_error_status(msg)
```

Comment:
hmm so we have this same stanza in three places (almost) repeated other than the msg having "export" or "open" - what do you think about adding another method for this logic called e.g. `_check_if_file_exists`? I think it would also make reading this function a bit clearer. We could also make the error message generic so we don't even need to pass in the operation: "Could not complete action on file {}. File does not exist"

Reply:
I like your `_check_if_file_exists` suggestion since we use it in three places: opening a file, preflight check, and exporting a file.

Reply:
actually thinking about this more, it would make more sense to verify the file system accurately reflects the local db during a sync rather than whenever the user tries to access the file. i'll modify the code so that we:

* check that the file exists and maybe use https://pypi.org/project/filelock/ so make sure it can't be accessed during open or export
* mark the file to be redownloaded if it's missing during a sync